### ISSUE TITLE
JS destructor not being called for classes

### ISSUE THREAD

**Issue by Zion-ICS** [swig/swig#329]
Steps to reproduce:
compile and install SWIG
run make check-examples
go to Examples/javascript/class
open runme.js and replace the top line with
var example = require("./build/Release/example");
Run the example

Note that it says that 2 shapes remain even after all the objects were destroyed.

We're running into the same issue using SWIG.
We need the destructor to be called because our destructors are cleaning up important state behind-the-scenes. It'd be a shame to have to provide interfaces to clean up everything just because the SWIG destructor isn't working.

If there were at least an interface to call the destructor, this bug wouldn't be as bad. But there's no delete_class function (note that python has new_class and delete_class functions).

**Comment by Zion-ICS** [comment]
I'm not sure whether the destructor for JavaScript was ever run on exit.
I just checked-out version 3.0.1, the first release with JavaScript examples.
I got the same result-- the destructor isn't being called.

**Comment by ojwb** [comment]
I don't think you can in general rely on the C++ destructor getting called on process exit for bindings in garbage collected target languages - that's not something SWIG can really control as it is down to the target language whether the proxy object gets destroyed (and finalised if that's a separate stage) in this situation, which is what's needed to call the C++ destructor.  It seems with node it doesn't.

Looking at `Examples/javascript/class/runme.js` that already explicitly does `delete c;` and `delete s;` yet reports (without the need for the change suggested above even):

> 2 shapes remain

So for node it seems the C++ destructor only gets triggered if/when GC runs and finalises the object after a `delete` on the JS side.

However adding an explicit `global.gc();` after the `delete` calls (which requires running with `node --expose-gc`) only changes that to:

> 1 shapes remain

So for some reason it seems the GC fails to reclaim both the objects here.

**Comment by ojwb** [comment]
Perhaps the way to resolve this then is to expose the existing `_wrap_delete_Shape` C++ function as `delete_Shape` in Javascript, (similar to what Python does apparently).
